pkg/gofr/datasource/file/ftp/file.go

```diff
@@ -0,0 +1,324 @@
+func (f *ftpFile) Close() error {
+	// Close the FTP response.
+	err := f.response.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Name returns the name of the file.
+func (f *ftpFile) Name() string {
+	return f.name
+}
+
+// Read reads data from the FTP file into the provided byte slice.
+func (f *ftpFile) Read(p []byte) (n int, err error) {
+	// Retrieve the file content from FTP server and read into p
+	r, err := f.conn.Retr(f.path)
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	for i := range b {
+		p[i] = b[i]
+	}
+
+	f.offset = int64(len(b))
+
+	return len(b), nil
+}
+
+// ReadAt reads data from the FTP file starting at the specified offset.
+func (f *ftpFile) ReadAt(p []byte, off int64) (n int, err error) {
+	// Retrieve file content from the specified offset
+	resp, err := f.conn.RetrFrom(f.path, uint64(off))
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(resp)
+	if err != nil {
+		return 0, err
+	}
+
+	// Deep Copy
+	for i := range b {
+		p[i] = b[i]
+	}
+
+	// setting current offset of the file
+	f.offset = off + int64(len(b))
+
+	return len(b), nil
+}
+
+// Seek sets the offset for the next Read or ReadAt operation.
+func (f *ftpFile) Seek(offset int64, whence int) (int64, error) {
+	var (
+		err error
+		r   ftpResponse
+	)
+
+	// Handle Seek operation based on whence parameter
+	r, err = f.conn.Retr(f.path)
+	if err != nil {
+		return 0, err
+	}
+
+	p, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	switch whence {
+	case io.SeekStart:
+		// Seek from the beginning of the file
+		if offset < 0 || offset > int64(len(p)) {
+			return 0, datasource.ErrOutOfRange
+		}
+
+		return offset, nil
+
+	case io.SeekEnd:
+		// Seek from the end of the file
+		if offset > 0 || offset < int64(-len(p)) {
+			return 0, datasource.ErrOutOfRange
+		}
+
+		return int64(len(p)) + offset, nil
+
+	case io.SeekCurrent:
+		// Seek from the current offset
+		if f.offset+offset >= int64(len(p)) || f.offset+offset < 0 {
+			return 0, datasource.ErrOutOfRange
+		}
+
+		return f.offset + offset, nil
+
+	default:
+	}
+
+	return 0, datasource.ErrOutOfRange
+}
+
+// Write writes data to the FTP file.
+func (f *ftpFile) Write(p []byte) (n int, err error) {
+	// Delete the existing file
+	err = f.conn.Delete(f.path)
+	if err != nil {
+		return 0, err
+	}
+	// Write new data to the Buffer
+	data := bytes.NewBuffer(p)
+	// Stor the file on ftp server
+	err = f.conn.Stor(f.path, data)
+	if err != nil {
+		return 0, err
+	}
+
+	f.offset = int64(len(p))
+
+	return len(p), nil
+}
+
+// WriteAt writes data to the FTP file starting at the specified offset.
+func (f *ftpFile) WriteAt(p []byte, off int64) (n int, err error) {
+	reader := bytes.NewReader(p)
+
+	// Write data to the FTP file from the specified offset
+	err = f.conn.StorFrom(f.path, reader, uint64(off))
+	if err != nil {
+		return 0, err
+	}
+
+	f.offset = off + int64(len(p))
```

Comment:
As with the `ReaderAt` above, `WriteAt` is also not supposed to update the underlying offset.